Add Tensor._make_wrapper_subclass

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #65340

I thought about a few possible ways of doing this.  The main hazard is
that if I create a CPU tensor that doesn't have any real storage, the
moment I actually try to access the data on the tensor I will segfault.
So I don't want to use _make_subclass on a "cpu meta tensor" because
the CPU meta tensor (with no subclass) is radioactive: printing it
will immediately cause a segfault.  So instead, I have to create
the CPU meta tensor AND subclass all in one go, and that means I need
another function for it.  One downside to doing it this way is
I need another overload for explicit strides, and in general it is
difficult to get the view relationships to all work out properly;
tracked at https://github.com/pytorch/pytorch/issues/65339

Fixes https://github.com/pytorch/pytorch/issues/62972
Fixes https://github.com/pytorch/pytorch/issues/62730

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D31057231](https://our.internmc.facebook.com/intern/diff/D31057231)

ezyang committed:
Add Tensor._make_wrapper_subclass

I thought about a few possible ways of doing this.  The main hazard is
that if I create a CPU tensor that doesn't have any real storage, the
moment I actually try to access the data on the tensor I will segfault.
So I don't want to use _make_subclass on a "cpu meta tensor" because
the CPU meta tensor (with no subclass) is radioactive: printing it
will immediately cause a segfault.  So instead, I have to create
the CPU meta tensor AND subclass all in one go, and that means I need
another function for it.  One downside to doing it this way is
I need another overload for explicit strides, and in general it is
difficult to get the view relationships to all work out properly;
tracked at https://github.com/pytorch/pytorch/issues/65339

Fixes https://github.com/pytorch/pytorch/issues/62972
Fixes https://github.com/pytorch/pytorch/issues/62730

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add Tensor._make_wrapper_subclass"


I thought about a few possible ways of doing this.  The main hazard is
that if I create a CPU tensor that doesn't have any real storage, the
moment I actually try to access the data on the tensor I will segfault.
So I don't want to use _make_subclass on a "cpu meta tensor" because
the CPU meta tensor (with no subclass) is radioactive: printing it
will immediately cause a segfault.  So instead, I have to create
the CPU meta tensor AND subclass all in one go, and that means I need
another function for it.  One downside to doing it this way is
I need another overload for explicit strides, and in general it is
difficult to get the view relationships to all work out properly;
tracked at https://github.com/pytorch/pytorch/issues/65339

Fixes https://github.com/pytorch/pytorch/issues/62972
Fixes https://github.com/pytorch/pytorch/issues/62730

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31057231).

albanD commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -364,6 +364,54 @@ static PyObject* THPVariable_make_subclass(PyObject* _ignored, PyObject* args, P
+  });
+  ParsedArgs<8> parsed_args{};
+  auto r = parser.parse(args, kwargs, parsed_args);
+  // TODO: handle has_torch_function, maybe?
```
There is no Tensor input here right?
That would be for Mode?

ezyang replied:
Yes. Also, if there's a hypothetical version of this which takes a Tensor and reads attributes directly from the Tensor maybe this should also be overrideable? I am not exactly sure what the use case for any of this is, it is mostly an argument from symmetry.

ezyang replied:
I guess technically torch function isn't getting a mode so we can't override this anyway lol

albanD commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -364,6 +364,54 @@ static PyObject* THPVariable_make_subclass(PyObject* _ignored, PyObject* args, P
+  // key, which will immediately segfault because the data pointer is null.  By
+  // forcing users to define __torch_dispatch__ we ensure this does not happen
+  if (PyObject_FastGetAttrString(cls, "__torch_dispatch__").ptr() == nullptr) {
+    throw torch::TypeError("%s must define __torch_dispatch__", ((PyTypeObject*)cls)->tp_name);
```
TORCH_CHECK

ezyang replied:
Is direct TypeError actually deprecated? There are still a lot of examples of this, some within this file.

albanD replied:
Ho, `TORCH_CHECK_TYPE()` to get a TypeError would work.
I think we want to make all of them consistent no?

albanD commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -364,6 +364,54 @@ static PyObject* THPVariable_make_subclass(PyObject* _ignored, PyObject* args, P
+
+  // don't bother releasing GIL here, as we are not allocating any nontrivial
+  // data
+  auto data = at::for_blob(nullptr, r.intlist(1))
```
Is there any doc anywhere about what for_blob is? That sounds very interesting?

ezyang replied:
yes go read ATen/templates/Functions.h

albanD reviewed (commented):
What are the properties from the Tensor created by for_blob?
Can we make sure that we just raise an error if we go below the Python key with it?

ezyang commented:
> What are the properties from the Tensor created by for_blob?

It's non-resizable (this may be a mistake, but we can fix it later), the data pointer is null, and we have no deleter for it. Or is there something else you're thinking of?

> Can we make sure that we just raise an error if we go below the Python key with it?

Not easily. We could add another dispatch key below Python to catch such cases (if I had spare dispatch keys), but I can't straightforwardly remove the CPU dispatch key as this will cause certain type tests to work incorrectly.

albanD commented:
> It's non-resizable (this may be a mistake, but we can fix it later), the data pointer is null, and we have no deleter for it. Or is there something else you're thinking of?

Sounds good!

albanD reviewed (approved):
LGTM!

zou3519 commented on test/test_python_dispatch.py:
```diff
@@ -33,10 +32,15 @@ class LoggingTensor(torch.Tensor):
-        # doesn't hold any memory (meta tensor is generally the preferred type
-        # of tensor you want to make a subclass from)...
-        r = torch.Tensor._make_subclass(cls, elem.to(WRAPPER_DEVICE), elem.requires_grad)
+        # The wrapping tensor (LoggingTensor) shouldn't hold any
+        # memory for the class in question, but it should still
+        # advertise the same device as before
+        r = torch.Tensor._make_wrapper_subclass(
+            cls, elem.size(),
+            # TODO: clone strides and storage aliasing
+            dtype=elem.dtype, layout=elem.layout,
+            device=elem.device, requires_grad=elem.requires_grad
+        )
```
What happens if someone calls .storage() on a LoggingTensor now?

ezyang replied:
You get a storage, but its data pointer is a null pointer and alas, you can probably induce a segfault this way. Good catch.

ezyang replied:
Indeed it segfaults

```
>>> class A(torch.Tensor):
...   def __torch_dispatch__(): pass
... 
>>> torch.Tensor._make_wrapper_subclass(A, (2,)).storage()[0]
Segmentation fault (core dumped)
```

I checked if meta tensors have this problem and they just error out when returning their storage, so I'll just do that here as a patch up

zou3519 commented on test/test_python_dispatch.py:
```diff
@@ -33,10 +32,15 @@ class LoggingTensor(torch.Tensor):
-        # doesn't hold any memory (meta tensor is generally the preferred type
-        # of tensor you want to make a subclass from)...
-        r = torch.Tensor._make_subclass(cls, elem.to(WRAPPER_DEVICE), elem.requires_grad)
+        # The wrapping tensor (LoggingTensor) shouldn't hold any
+        # memory for the class in question, but it should still
+        # advertise the same device as before
```
Add test?

zou3519 reviewed (commented):
This is cool! Will follow along with strides and view relationship handling (the lack of view relationship handling has been causing internal asserts with functorch on debug builds)

Lezcano commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -364,6 +364,54 @@ static PyObject* THPVariable_make_subclass(PyObject* _ignored, PyObject* args, P
   END_HANDLE_TH_ERRORS
 }
 
+static PyObject* THPVariable_make_wrapper_subclass(PyObject* _ignored, PyObject* args, PyObject* kwargs) {
```
nit. Write it as follows to avoid unused variable warnings
```suggestion
static PyObject* THPVariable_make_wrapper_subclass(PyObject* , PyObject* args, PyObject* kwargs) {
```

ezyang committed:
Update on "Add Tensor._make_wrapper_subclass"


I thought about a few possible ways of doing this.  The main hazard is
that if I create a CPU tensor that doesn't have any real storage, the
moment I actually try to access the data on the tensor I will segfault.
So I don't want to use _make_subclass on a "cpu meta tensor" because
the CPU meta tensor (with no subclass) is radioactive: printing it
will immediately cause a segfault.  So instead, I have to create
the CPU meta tensor AND subclass all in one go, and that means I need
another function for it.  One downside to doing it this way is
I need another overload for explicit strides, and in general it is
difficult to get the view relationships to all work out properly;
tracked at https://github.com/pytorch/pytorch/issues/65339

Fixes https://github.com/pytorch/pytorch/issues/62972
Fixes https://github.com/pytorch/pytorch/issues/62730

Signed-off-by: Edward Z. Yang <ezyangfb.com>

Differential Revision: [D31057231](https://our.internmc.facebook.com/intern/diff/D31057231)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31057231).

anjali411 commented on torch/csrc/DynamicTypes.cpp:
```diff
@@ -63,6 +63,9 @@ PyTypeObject* getPyTypeObject(
   if (storage.device_type() == at::DeviceType::Meta) {
     TORCH_CHECK_NOT_IMPLEMENTED(false, "python bindings for meta storage objects not supported");
   }
+  if (storage.data() == nullptr && storage.nbytes() != 0) {
```
we could instead set `set_storage_access_should_throw` https://github.com/pytorch/pytorch/blob/760aefd34d4cb2edb6fb11eea4e72e2d62b51498/c10/core/TensorImpl.h#L2342-L2344 in `THPVariable_make_wrapper_subclass`

ezyang replied:
The thing is, per https://github.com/pytorch/pytorch/issues/65339 in the long term we do want storage represented adequately
